proposal/TommyLin/README.rst

```diff
@@ -0,0 +1,256 @@
+
+.. [#] Control could be PC or embedded system such as Raspberry PI.
+.. [#] This can be achived by using `Canon Digital Camera Software Development Kit (CD-SDK) <https://asia.canon/en/campaign/developerresources>`_.
+
+
+**Focus stacking** (software)
+_____________________________
+
+[**Scan for images**] ==> [**Grouping**] ==> [**Edge detect**] ==> [*Aligment* [#]_] ==> [*White Balance*] ==> [**Focus stacking**]
+
+.. [#] Aligment and white balance are not included in this project. Tested images are created and always aligned and white balanced.
+
+API Description
+===============
+
+User APIs: (Open to users)
+__________________________
+
+.. [focus_stacking] Auto scan mode: Users select specific directory that contain photos. This API
+  will scan directory and determinate the groups of photos depend on SSIM value. Then stack these
+  images together within one group.
+
+int focus_stacking(path target_path, fd* image_fd)
+  Do focus stacking in specified folder
+
+Return value:
+  0: Success
+
+  Others: Fail
+
+target_path:
+  Folder to store photos
+
+image_fd:
+  Photo file descriptors
+
+
+Private APIs (Internal used only)
+________________________________________
+
+*image gaussian_blur(image input, kernel gaussian)*
+
+*image edge_detect(image input, kernel laplacian)*
+
+*image stacking(image_ptr source, image_ptr log_mask)*
+
+*int ssim(image image1, image image2)*
+
+
+Types and Strcutures
+====================
+
+::
+
+  struct image_t {
+      int   hight;
+      int   width;
+      char** raw_data;
+  }
+
+::
+
+  struct kernel_t {
+      int   rank;
+      char* kernel;
+  }
+
+
+Engineering Infrastructure
+==========================
+
+Algorism Performance Evaluation
+_______________________________
+
+To evaluate the algorism performance in this project, testing images are created 
+from a source image(focused). This also prevent aligment procedure to do focus
+stacking. Here is the steps to evaluate algorism performance.
+
+1. Get a clear and focused source image (image1)
+2. Smooth souce image with a set of mask by Gaussian blur kernel
+3. Through the focus stacking API [focus_stacking]_ to combine these blurred images into image2
+4. Calculate SSIM of image1 and image2 to tell the performance of focus stacking algorism.
+
+Coding Style Check
+__________________
+
+checkpatch.pl
+
+::
+
+    $ $(kernel)/scripts/checkpatch.pl -f source.cpp
+
+Static Analysis Tool
+____________________
+
+`cppcheck <http://cppcheck.sourceforge.net/>`_
+
+A static analysis tool for C/C++ code
+
+
+
+::
+
+    $ cppcheck .
+
+
+Schedule
+========
+
+* Week 1: [Python] Generate defocused photos (Gaussian blur)
+* Week 2: [C++] Evaluate SSIM of two photos
+* Week 3: [Python] Setup testing environment
+* Week 4: [Python] Browse directory and read image
+* Week 5: [C++] Gaussian blur & Laplacian edge detection
+* Week 6: [C++] Image stacking
+* Week 7: [Python/C++] Debug and optimization
+* Week 8: Github CI
```

Comment:
It's kind of missing the point of having CI at the last stage of development.  Will your code have automatic testing before this stage?

Reply:
There will be testing scripts or procedures lauched manually in the early days. Github CI job putting in week 8 was in case I have no time to make it happen. If the task in week 3 was done warly, I will try Github CI in that time.

Reply:
I would suggest to move CI to week 3 so that it provides you the best productivity.  In my experience, the later to have a CI, the slower the development becomes.